src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs

```diff
@@ -1,161 +1,132 @@
-            }
-
-            return new OkResponse<List<PlaceholderNameAndValue>>(placeHolderNameAndValues);
+                new($"{PlaceHolderLeft}{catchAllQueryPlaceholder}{PlaceHolderRight}", string.Empty),
+            });
         }
 
-        private static bool NoMoreForwardSlash(string pathTemplate, int counterForTemplate)
-        {
-            return !pathTemplate.Substring(counterForTemplate).Contains("/");
-        }
+        // Find matching groups from path and query
+        IList<Group> groups = FindGroups(path, query, pathTemplate);
+        List<PlaceholderNameAndValue> placeholderNameAndValues = groups
+            .Select(group => new PlaceholderNameAndValue($"{PlaceHolderLeft}{group.Name}{PlaceHolderRight}", group.Value))
+            .ToList();
 
-        private static bool NotPassedQueryString(string pathTemplate, int counterForTemplate)
-        {
-            return !pathTemplate.Substring(0, counterForTemplate).Contains('?');
-        }
-
-        private static bool PassedQueryString(string pathTemplate, int counterForTemplate)
-        {
-            return pathTemplate.Substring(0, counterForTemplate).Contains('?');
-        }
-
-        private static bool IsCatchAll(string path, int counterForPath, string pathTemplate)
-        {
-            return string.IsNullOrEmpty(path) || (path.Length > counterForPath && path[counterForPath] == '/') && pathTemplate.Length > 1
-                     && pathTemplate.Substring(0, 2) == "/{"
-                     && pathTemplate.IndexOf('}') == pathTemplate.Length - 1;
-        }
-
-        private static bool IsCatchAllAfterOtherPlaceholders(string pathTemplate, int counterForTemplate)
-            => (pathTemplate[counterForTemplate] == '/' || pathTemplate[counterForTemplate] == '?')
-                && (counterForTemplate < pathTemplate.Length - 1)
-                && (pathTemplate[counterForTemplate + 1] == '{')
-                && NoMoreForwardSlash(pathTemplate, counterForTemplate + 1);
+        return new OkResponse<List<PlaceholderNameAndValue>>(placeholderNameAndValues);
+    }
 
-        private static bool NothingAfterFirstForwardSlash(string path)
+    /// <summary>
+    /// Finds the placeholders in the request path and query.
+    /// We use a regex pattern to match the placeholders in the path template.
+    /// We have two slight optimizations:
+    /// - First, we skip the query if it is not present in the path template.
+    /// - Second, we append a trailing slash to the path if it is a catch-all path.
+    /// </summary>
+    /// <param name="path">The request path.</param>
+    /// <param name="query">The query parameters.</param>
+    /// <param name="template">The path template.</param>
+    /// <returns>The matching groups.</returns>
+    private static IList<Group> FindGroups(string path, string query, string template)
+    {
+        Regex regex = new($@"\{PlaceHolderLeft}(.*?)\{PlaceHolderRight}", RegexOptions.None, TimeSpan.FromSeconds(1));
```

Comment:
I just don't get it. Why can't you write a constant `Regex` pattern string? Don't know how to escape special characters? 😄 
```suggestion
        Regex regex = new(@"\{(.*?)\}", RegexOptions.None, TimeSpan.FromSeconds(1));
```
Also keep in mind PR #1348 which I'm working now

Reply:
@raman-m wtf! That's a bit rude. As I wrote you earlier, we might want to use other characters as placeholders...

Reply:
Maybe I misunderstood the problem... Doesn't const pattern solve the problem?.. and we have to use interpolation?
Why are you so emotional today? Your client f...ed you? ))

Reply:
No no, it's fine, let's do this like that, it's sound for me! And besides, string interpolation has an impact on the performance, we don't want that @raman-m

Reply:
Yes, we are performance funs. 😉